vertical-pod-autoscaler/enhancements/3913-customized-recomender-vpa.md

```diff
@@ -0,0 +1,272 @@
+
+## Motivation
+
+A VPA is used to resize containers in pods when the actual CPU/memory usage of a container 
+is significantly different from the resources requested. Resource usage-based recommendation 
+is the basic approach that resizes containers according to the actual usage observed and is 
+implemented in the default VPA recommender. Users can configure the time window and a certain 
+percentile of observed usage in the past as the prediction of future requests/limits for CPU/memory.
+
+However, as containers running different types of workloads may have different resource usage patterns, 
+there is no universal policy that applies to all. The existing VPA recommender may not accurately 
+predict future resource usage when containers exhibit certain resource usage behaviors, 
+such as trending, periodically changing, or occasional spikes, resulting in significant 
+over-provisioning and OOM kills for microservices. Learning different types of resource usage 
+behaviors for workloads and applying different algorithms to improve resource utilization 
+(CPU and Memory) predictions can significantly reduce over-provisioning and OOM kills in VPA.
+
+### Goals
+
+- Allow the VPA object to specify a customized recommender to use. 
+- Allow the VPA object to use the default recommender when no recommender is specified.
+
+### Non-Goals
+
+- We assume no pod uses two recommenders at the same time.
+- We do not resolve conflicts between recommenders.
+
+## Proposal
+
+### User Stories
+
+#### Story 1 
+
+- Containers with Cyclic Patterns in Resource Usage
+
+Containers used in monitoring may receive load periodically to process but need to be long-running 
+to listen to incoming traffic. Thus, these containers usually exhibit cyclic patterns, alternating 
+between usage spikes and idling. Resizing containers according to usage observed in the previous 
+time window may always lead to under-provision for a short period when the load spikes just arrive. 
+The problem will even lead to frequent pod evictions when the pod was resized according to the idling 
+period and the host resource has been taken by other pods.
+
+#### Story 2 
+
+- Containers with Different but Recurrent Behaviors in Resource Usage
+
+Containers running spark/deep learning training workloads are known to show recurring and repeating 
+patterns in resource usage. Prior research has shown that different but recurrent behaviors occur 
+for different containerized tasks, such as Spark or deep learning training. These common patterns can 
+be represented by phases, which display similar resource usage of computational resources over time. 
+There are common sequences of patterns for different executions of the workload and they can be used 
+to proactively predict future resource usage more accurately. The default recommender in the current 
+VPA adopts a reactive approach so a more proactive recommender is needed for these types of workload.
+
+### Implementation Details
```

Comment:
I think it's clear from the context, especially from rejected alternatives, but it still would be good to explicitly state that the customised recommenders are expected to be running outside VPA binary (or at least that they can run separately).

Maybe an image/description showing how VPA controller would look like (multiple recommender binaries, still a single Updater & web-hook)?